python-313/typing/readonly.py

```diff
@@ -0,0 +1,48 @@
+"""Demonstration of PEP 705: TypedDict: read-only items
+
+Use PyLance in VS Code by setting Python › Analysis: Type Checking Mode or run
+the Pyright CLI:
+
+    $ python -m pip install pyright $ pyright --pythonversion 3.13 .
+
+Extension of TypedDict:
+https://realpython.com/python38-new-features/#more-precise-types
+"""
+
+from typing import NotRequired, ReadOnly, TypedDict
+
+# class Version(TypedDict):
+#     version: str
+#     release_year: NotRequired[int | None]
+
+
+# class PythonVersion(TypedDict):
+#     version: str
+#     release_year: int
+
+
+class Version(TypedDict):
+    version: ReadOnly[str]
+    release_year: ReadOnly[NotRequired[int | None]]
+
+
+class PythonVersion(TypedDict):
+    version: ReadOnly[str]
+    release_year: ReadOnly[int]
+
+
+py313 = PythonVersion(version="3.13", release_year=2024)
+
+# Alternative syntax, using TypedDict as an annotation
+# py313: PythonVersion = {"version": "3.13", "release_year": 2024}
+
+
+def get_version_info(ver: Version) -> str:
+    if "release_year" in ver:
+        return f"Version {ver['version']} released in {ver['release_year']}"
+    else:
+        return f"Version {ver['version']}"
+
+
+# Only allowed to use PythonVersion instead of Version if the fields are ReadOnly
+print(get_version_info(py313))
```

Comment:
I find this counter-intuitive since the `Version` and `PythonVersion` classes specify separate types that don't belong to the same type hierarchy. They share the same structure, so they're essentially the same type by means of static duck typing, but they don't explicitly define a protocol. I guess, both are dicts with the same set of key-value pairs, so that should be enough for the type checker to treat them as equals.

Reply:
Right, I believe that TypedDict is all about structural typing. The [PEP](https://peps.python.org/pep-0705/) talks a bit about how the capabilities of `ReadOnly` can be done with some complicated use of protocols instead.

I haven't quite internalized the rules for when one TypedDict can pass as the other, but everytime I sit down and think for a while, it seems to add up :see_no_evil: